src/main/java/net/fabricmc/tinyremapper/TinyRemapper.java

```diff
@@ -408,7 +401,11 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
 				@Override
 				public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
 					if (file.toString().endsWith(".class")) {
-						ClassInstance res = analyze(isInput, tags, srcPath, Files.readAllBytes(file));
+						OptionalInt mrjVersion = OptionalInt.empty();
```

Comment:
Instead of using file start with check, you should pull it out so you walk the META-INF and other directories separately

Reply:
I do not see the advantage of walk the META-INF separately. And the current way is much easier to implement.

Reply:
This current way is bad performance-wise as it tries to parse the version multiple times while a walk within the META-INF can find all versions at once and avoid creating OptionalInt instances repeatedly.

Reply:
Yes, but that will increase the complexity of code (and I don't think the current code will create too many additional OptionalInt anyway). I will leave it for now. My current priority is to fix the class dependency resolution (i.e. using the correct version instead of using the Java 8 version for everything).

Reply:
Well, that doesn't. That actually makes things clearer as you already have to build a class graph for each release present, and splitting the version-specific and common release file visit code will increase clarity and make code more maintainable in the future, not to mention the performance boost it is already bringing.